src/main/java/seedu/uninurse/model/PersistentUninurseBook.java

```diff
@@ -0,0 +1,89 @@
+package seedu.uninurse.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import seedu.uninurse.commons.core.Config;
+import seedu.uninurse.logic.commands.CommandResult;
+
+/**
+ * Versions of UninurseBookSnapshot after every command.
+ */
+public class PersistentUninurseBook extends UninurseBook {
```

Comment:
I think having a `PersistentUninurseBook` inherit from `UninurseBook` doesn't violate LSP, since `PersistentUninurseBook` is essentially a `UninurseBook`, but it additionally supports undo and redo operations "somehow". The implementation of how it achieves this (i.e. the `List<UninurseBook>`) is irrelevant.

Also, I think allowing `PersistentUninurseBook` to store other `PersistentUninurseBook`s doesn't change or break the functionality.

Thoughts?

Reply:
Since `PersistentUninurseBook` and `UninurseBookSnapshot` inherits from `UninurseBook`, right now they both have all the functionalities of `UninurseBook`.

Since `PersistentUninurseBook` stores a list of `UninurseBookSnapshot`, it means that if I call `PersistentUninurseBook#undo` a few times and modify it (e.g. using `PersistentUninurseBook#setPersons`) and call `PersistentUninurseBook#redo` again the same amount of times, the modification I made would have been nullified according to your implementation (because `PersistentUninurseBook#makeSnapshot` was not invoked).

However that should not be the case since `UninurseBookSnapshot` is a `UninurseBook`, which means that it is its own `UninurseBook` and the modification should not have been nullified. The reason why I said LSP is violated is because according to the abstractions that `PersistentUninurseBook` and `UninurseBookSnapshot` are indeed "`UninurseBook`", this particular scenario violates the behavior that `UninurseBookSnapshot` functions like a `UninurseBook` since the result from invoking any methods from `UninurseBookSnapshot` that were inherited from `UninurseBook` cannot be guaranteed. This directly implies that in all possible scenarios that can be constructed using these 3 classes, there contain scenarios where `UninurseBookSnapshot` cannot replace `UninurseBook`, which is a violation of LSP. A slightly more complicated scenario can be constructed for `PersistentUninurseBook`.

This is why I believe that `PersistentUninurseBook` and `UninurseBookSnapshot` should be some type of association/container class that has a copy of the `UninurseBook` rather than being `UninurseBook`s themselves. 

`UninurseBookSnapshot` also violates ISP as it never has to use the methods of `UninurseBook` except `UninurseBook#getPersonList`

Although I recognize that your implementation does indeed prevent unintended behavior and thus functionality is not affected, however I have to point out that the implementation and the abstractions it is based on is inconsistent and the fact that you might need to explain why your implementation does not invoke unintended behavior even though the abstractions are as such is an issue in itself. This also violates DIP as now your details of your implementation does not depend on abstractions, which makes the abstraction that `PersistentUninurseBook` and `UninurseBookSnapshot` inheriting from `UninurseBook` pointless.

If you still think otherwise I don't mind spending time discussing this further but hopefully what I say still makes some sense to you.

Reply:
I agree with your reasoning regarding `UninurseBookSnapshot`. I still believe that `PersistentUninurseBook` inheriting from `UninurseBook` doesn't break LSP, though; wouldn't having called `PersistentUninurseBook#undo` implies that the object must have been a `PersistentUninurseBook`? 

I'd like to discuss this further, but in the interest of time, I've updated the implementation of `UninurseBookSnapshot` and `PersistentUninurseBook` as a wrapper class according to your suggestions.